diff --git a/target/linux/ar71xx/base-files/etc/board.d/02_network b/target/linux/ar71xx/base-files/etc/board.d/02_network
index 13a224a6b..912f87200 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/02_network
+++ b/target/linux/ar71xx/base-files/etc/board.d/02_network
@@ -178,6 +178,7 @@ ar71xx_setup_interfaces()
 	tl-wr720n-v3|\
 	tl-wr810n|\
 	tl-wr810n-v2|\
+	tl-wr800n-v2|\
 	wpe72|\
 	wrtnode2q)
 		ucidef_set_interfaces_lan_wan "eth1" "eth0"
diff --git a/target/linux/ar71xx/base-files/etc/diag.sh b/target/linux/ar71xx/base-files/etc/diag.sh
index 666c26352..ddc99b1ea 100644
--- a/target/linux/ar71xx/base-files/etc/diag.sh
+++ b/target/linux/ar71xx/base-files/etc/diag.sh
@@ -487,6 +487,9 @@ get_status_led() {
 	tl-wr941nd-v5)
 		status_led="tp-link:green:system"
 		;;
+	tl-wr800n-v2)
+		status_led="tp-link:green:qss"
+		;;
 	tl-wr841n-v9)
 		status_led="tp-link:green:qss"
 		;;
diff --git a/target/linux/ar71xx/base-files/lib/ar71xx.sh b/target/linux/ar71xx/base-files/lib/ar71xx.sh
index 43c501ebf..c871c3c06 100755
--- a/target/linux/ar71xx/base-files/lib/ar71xx.sh
+++ b/target/linux/ar71xx/base-files/lib/ar71xx.sh
@@ -1391,6 +1391,9 @@ ar71xx_board_detect() {
 	*"TL-WR841N/ND v8")
 		name="tl-wr841n-v8"
 		;;
+	*"TL-WR800N v2")
+		name="tl-wr800n-v2"
+		;;
 	*"TL-WR841N/ND v9")
 		name="tl-wr841n-v9"
 		;;
diff --git a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
index bee5f4811..e4d64cb10 100755
--- a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
@@ -485,6 +485,7 @@ platform_check_image() {
 	tl-wr841n-v7|\
 	tl-wr841n-v8|\
 	tl-wr841n-v9|\
+	tl-wr800n-v2|\
 	tl-wr841n-v11|\
 	tl-wr842n-v2|\
 	tl-wr842n-v3|\
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/Makefile b/target/linux/ar71xx/files/arch/mips/ath79/Makefile
index e13e748d5..4b2d33cbd 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/Makefile
+++ b/target/linux/ar71xx/files/arch/mips/ath79/Makefile
@@ -243,6 +243,7 @@ obj-$(CONFIG_ATH79_MACH_TL_WR802N_V1)		+= mach-tl-wr802n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR802N_V2)		+= mach-tl-wr802n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR810N)		+= mach-tl-wr810n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR810N_V2)		+= mach-tl-wr810n.o
+obj-$(CONFIG_ATH79_MACH_TL_WR800N_V2)		+= mach-tl-wr800n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR840N_V2)		+= mach-tl-wr841n-v9.o
 obj-$(CONFIG_ATH79_MACH_TL_WR841N_V1)		+= mach-tl-wr841n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR841N_V8)		+= mach-tl-wr841n-v8.o
diff --git a/target/linux/ar71xx/image/tiny-tp-link.mk b/target/linux/ar71xx/image/tiny-tp-link.mk
index 5874808d2..52f417626 100644
--- a/target/linux/ar71xx/image/tiny-tp-link.mk
+++ b/target/linux/ar71xx/image/tiny-tp-link.mk
@@ -693,3 +693,12 @@ define Device/tl-wr941nd-v6-cn
   TPLINK_HWID := 0x09410006
 endef
 TARGET_DEVICES += tl-wr941nd-v6-cn
+
+define Device/tl-wr800n-v2
+  $(Device/tplink-8mlzma)
+  DEVICE_TITLE := TP-LINK TL-WR800N v2
+  BOARDNAME := TL-WR800N-v2
+  DEVICE_PROFILE := TLWR800
+  TPLINK_HWID := 0x08000002
+endef
+TARGET_DEVICES += tl-wr800n-v2
diff --git a/target/linux/ath79/image/tiny-tp-link.mk b/target/linux/ath79/image/tiny-tp-link.mk
index c918c6baa..82a827b21 100644
--- a/target/linux/ath79/image/tiny-tp-link.mk
+++ b/target/linux/ath79/image/tiny-tp-link.mk
@@ -486,3 +486,13 @@ define Device/tplink_tl-wr941n-v7-cn
   TPLINK_HWID := 0x09410007
 endef
 TARGET_DEVICES += tplink_tl-wr941n-v7-cn
+
+define Device/tplink_tl-wr800n-v2
+  $(Device/tplink-16mlzma)
+  SOC := qca9533
+  DEVICE_MODEL := TL-WR800N
+  DEVICE_VARIANT := v2
+  TPLINK_HWID := 0x08000002
+endef
+TARGET_DEVICES += tplink_tl-wr800n-v2
+
