diff --git a/target/linux/ar71xx/base-files/etc/board.d/01_leds b/target/linux/ar71xx/base-files/etc/board.d/01_leds
index d84150347..19d4fd57d 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/01_leds
+++ b/target/linux/ar71xx/base-files/etc/board.d/01_leds
@@ -794,11 +794,11 @@ tl-mr3420)
 	;;
 tl-mr3220-v2|\
 tl-wr741nd-v4)
-	ucidef_set_led_netdev "wan" "WAN" "tp-link:green:wan" "eth1"
+	ucidef_set_led_netdev "wan" "WAN" "tp-link:green:lan4" "eth1"
 	ucidef_set_led_switch "lan1" "LAN1" "tp-link:green:lan1" "switch0" "0x04"
 	ucidef_set_led_switch "lan2" "LAN2" "tp-link:green:lan2" "switch0" "0x08"
 	ucidef_set_led_switch "lan3" "LAN3" "tp-link:green:lan3" "switch0" "0x10"
-	ucidef_set_led_switch "lan4" "LAN4" "tp-link:green:lan4" "switch0" "0x02"
+	ucidef_set_led_switch "lan4" "LAN4" "tp-link:green:wan" "switch0" "0x02"
 	ucidef_set_led_wlan "wlan" "WLAN" "tp-link:green:wlan" "phy0tpt"
 
 	case "$board" in
@@ -942,6 +942,7 @@ tl-wr902ac-v1)
 	ucidef_set_led_wlan "wlan2g" "WLAN2G" "$board:green:wlan2g" "phy1tpt"
 	;;
 tl-wr940n-v4|\
+tl-wr882n-v1|\
 tl-wr941nd-v6)
 	ucidef_set_led_netdev "wan" "WAN" "tp-link:blue:wan" "eth0"
 	ucidef_set_led_switch "lan1" "LAN1" "tp-link:blue:lan1" "switch0" "0x10"
diff --git a/target/linux/ar71xx/base-files/etc/board.d/02_network b/target/linux/ar71xx/base-files/etc/board.d/02_network
index 13a224a6b..39a9f493c 100755
--- a/target/linux/ar71xx/base-files/etc/board.d/02_network
+++ b/target/linux/ar71xx/base-files/etc/board.d/02_network
@@ -178,6 +178,7 @@ ar71xx_setup_interfaces()
 	tl-wr720n-v3|\
 	tl-wr810n|\
 	tl-wr810n-v2|\
+	tl-wr800n-v2|\
 	wpe72|\
 	wrtnode2q)
 		ucidef_set_interfaces_lan_wan "eth1" "eth0"
@@ -502,10 +503,11 @@ ar71xx_setup_interfaces()
 	tl-wr841n-v8|\
 	tl-wr842n-v2|\
 	tl-wr941nd-v5|\
+	tl-wr882n-v1|\
 	tl-wr942n-v1)
 		ucidef_set_interfaces_lan_wan "eth1.1" "eth0"
-		ucidef_add_switch "switch0" \
-			"0@eth1" "1:lan:4" "2:lan:1" "3:lan:2" "4:lan:3"
+                ucidef_add_switch "switch0" \
+                        "0@eth1" "1:lan:1" "2:lan:2" "3:lan:3" "4:lan:4"
 		;;
 	archer-c7-v4|\
 	archer-c7-v5|\
@@ -530,7 +532,7 @@ ar71xx_setup_interfaces()
 	tl-wr741nd-v4)
 		ucidef_set_interfaces_lan_wan "eth0.1" "eth1"
 		ucidef_add_switch "switch0" \
-			"0@eth0" "1:lan:4" "2:lan:1" "3:lan:2" "4:lan:3"
+			"0@eth0" "1:lan:1" "2:lan:2" "3:lan:3" "4:lan:4"
 		;;
 	tl-wr841n-v1|\
 	tl-wr941nd)
diff --git a/target/linux/ar71xx/base-files/etc/diag.sh b/target/linux/ar71xx/base-files/etc/diag.sh
index 666c26352..15a22c7ea 100644
--- a/target/linux/ar71xx/base-files/etc/diag.sh
+++ b/target/linux/ar71xx/base-files/etc/diag.sh
@@ -487,12 +487,18 @@ get_status_led() {
 	tl-wr941nd-v5)
 		status_led="tp-link:green:system"
 		;;
+	tl-wr800n-v2)
+		status_led="tp-link:green:qss"
+		;;
 	tl-wr841n-v9)
 		status_led="tp-link:green:qss"
 		;;
 	tl-wr940n-v6)
 		status_led="tp-link:orange:diag"
 		;;
+	tl-wr882n-v1)
+		status_led="tp-link:white:status"
+		;;
 	tl-wdr6500-v2)
 		status_led="tp-link:white:system"
 		;;
diff --git a/target/linux/ar71xx/base-files/lib/ar71xx.sh b/target/linux/ar71xx/base-files/lib/ar71xx.sh
index 43c501ebf..fb1df81ca 100755
--- a/target/linux/ar71xx/base-files/lib/ar71xx.sh
+++ b/target/linux/ar71xx/base-files/lib/ar71xx.sh
@@ -276,6 +276,9 @@ tplink_board_detect() {
 	"090100"*)
 		model="TP-Link TL-WA901N/ND"
 		;;
+	"088200"*)
+		model="TP-Link TL-WR882N"
+		;;
 	"094000"*)
 		model="TP-Link TL-WR940N"
 		;;
@@ -1391,6 +1394,9 @@ ar71xx_board_detect() {
 	*"TL-WR841N/ND v8")
 		name="tl-wr841n-v8"
 		;;
+	*"TL-WR800N v2")
+		name="tl-wr800n-v2"
+		;;
 	*"TL-WR841N/ND v9")
 		name="tl-wr841n-v9"
 		;;
@@ -1403,6 +1409,9 @@ ar71xx_board_detect() {
 	*"TL-WR902AC v1")
 		name="tl-wr902ac-v1"
 		;;
+	*"TL-WR882N v1")
+		name="tl-wr882n-v1"
+		;;
 	*"TL-WR940N v4")
 		name="tl-wr940n-v4"
 		;;
diff --git a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
index bee5f4811..e8460cb4c 100755
--- a/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ar71xx/base-files/lib/upgrade/platform.sh
@@ -485,10 +485,12 @@ platform_check_image() {
 	tl-wr841n-v7|\
 	tl-wr841n-v8|\
 	tl-wr841n-v9|\
+	tl-wr800n-v2|\
 	tl-wr841n-v11|\
 	tl-wr842n-v2|\
 	tl-wr842n-v3|\
 	tl-wr902ac-v1|\
+	tl-wr882n-v1|\
 	tl-wr940n-v4|\
 	tl-wr940n-v6|\
 	tl-wr941nd|\
diff --git a/target/linux/ar71xx/config-4.14 b/target/linux/ar71xx/config-4.14
index 0f3f2c71b..7d03cd21f 100644
--- a/target/linux/ar71xx/config-4.14
+++ b/target/linux/ar71xx/config-4.14
@@ -238,6 +238,7 @@ CONFIG_ATH79=y
 # CONFIG_ATH79_MACH_TL_WR841N_V8 is not set
 # CONFIG_ATH79_MACH_TL_WR841N_V9 is not set
 # CONFIG_ATH79_MACH_TL_WR902AC_V1 is not set
+# CONFIG_ATH79_MACH_TL_WR882N_V1 is not set
 # CONFIG_ATH79_MACH_TL_WR940N_V4 is not set
 # CONFIG_ATH79_MACH_TL_WR941ND is not set
 # CONFIG_ATH79_MACH_TL_WR941ND_V6 is not set
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt b/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
index 0e4dae75e..cbcc2e7ba 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
+++ b/target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt
@@ -2018,6 +2018,15 @@ config ATH79_MACH_TL_WR940N_V4
 	select ATH79_DEV_M25P80
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_TL_WR882N_V1
+	bool "TP-LINK TL-WR882N v1 support"
+	select SOC_QCA956X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_WMAC
+
 config ATH79_MACH_TL_WR942N_V1
 	bool "TP-LINK TL-WR942N v1 support"
 	select SOC_QCA956X
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/Makefile b/target/linux/ar71xx/files/arch/mips/ath79/Makefile
index e13e748d5..53906c664 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/Makefile
+++ b/target/linux/ar71xx/files/arch/mips/ath79/Makefile
@@ -243,6 +243,7 @@ obj-$(CONFIG_ATH79_MACH_TL_WR802N_V1)		+= mach-tl-wr802n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR802N_V2)		+= mach-tl-wr802n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR810N)		+= mach-tl-wr810n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR810N_V2)		+= mach-tl-wr810n.o
+obj-$(CONFIG_ATH79_MACH_TL_WR800N_V2)		+= mach-tl-wr800n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR840N_V2)		+= mach-tl-wr841n-v9.o
 obj-$(CONFIG_ATH79_MACH_TL_WR841N_V1)		+= mach-tl-wr841n.o
 obj-$(CONFIG_ATH79_MACH_TL_WR841N_V8)		+= mach-tl-wr841n-v8.o
@@ -250,6 +251,7 @@ obj-$(CONFIG_ATH79_MACH_TL_WR841N_V9)		+= mach-tl-wr841n-v9.o
 obj-$(CONFIG_ATH79_MACH_TL_WR902AC_V1)		+= mach-tl-wr902ac-v1.o
 obj-$(CONFIG_ATH79_MACH_TL_WR941ND)		+= mach-tl-wr941nd.o
 obj-$(CONFIG_ATH79_MACH_TL_WR941ND_V6)		+= mach-tl-wr941nd-v6.o
+obj-$(CONFIG_ATH79_MACH_TL_WR882N_V1)		+= mach-tl-wr882n-v1.o
 obj-$(CONFIG_ATH79_MACH_TL_WR940N_V4)		+= mach-tl-wr940n-v4.o
 obj-$(CONFIG_ATH79_MACH_TL_WR942N_V1)		+= mach-tl-wr942n-v1.o
 obj-$(CONFIG_ATH79_MACH_TS_D084)		+= mach-ts-d084.o
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr741nd-v4.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr741nd-v4.c
index 851b7624f..bd8003481 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr741nd-v4.c
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr741nd-v4.c
@@ -54,8 +54,8 @@ static struct flash_platform_data tl_wr741ndv4_flash_data = {
 
 static struct gpio_led tl_wr741ndv4_leds_gpio[] __initdata = {
 	{
-		.name		= "tp-link:green:lan1",
-		.gpio		= TL_WR741NDV4_GPIO_LED_LAN1,
+		.name		= "tp-link:green:wan",
+		.gpio		= TL_WR741NDV4_GPIO_LED_WAN,
 		.active_low	= 0,
 	}, {
 		.name		= "tp-link:green:lan2",
@@ -78,8 +78,8 @@ static struct gpio_led tl_wr741ndv4_leds_gpio[] __initdata = {
 		.gpio		= TL_WR741NDV4_GPIO_LED_SYSTEM,
 		.active_low	= 1,
 	}, {
-		.name		= "tp-link:green:wan",
-		.gpio		= TL_WR741NDV4_GPIO_LED_WAN,
+		.name		= "tp-link:green:lan1",
+		.gpio		= TL_WR741NDV4_GPIO_LED_LAN1,
 		.active_low	= 0,
 	}, {
 		.name		= "tp-link:green:wlan",
@@ -134,7 +134,7 @@ static void __init tl_ap121_setup(void)
 	u8 *mac = (u8 *) KSEG1ADDR(0x1f01fc00);
 	u8 *ee = (u8 *) KSEG1ADDR(0x1fff1000);
 
-	ath79_setup_ar933x_phy4_switch(true, true);
+	ath79_setup_ar933x_phy4_switch(false,false);
 
 	ath79_gpio_function_disable(AR933X_GPIO_FUNC_ETH_SWITCH_LED0_EN |
 				    AR933X_GPIO_FUNC_ETH_SWITCH_LED1_EN |
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr800n.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr800n.c
new file mode 100644
index 000000000..00efa9e34
--- /dev/null
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr800n.c
@@ -0,0 +1,149 @@
+/*
+ * TP-LINK TL-WR810N board support
+ *
+ * Copyright (c) 2012 Qualcomm Atheros
+ * Copyright (c) 2012 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (c) 2016 Jens Steinhauser <jens.steinhauser@gmail.com>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <linux/gpio.h>
+#include <linux/platform_device.h>
+#include <linux/ar8216_platform.h>
+
+#include <asm/mach-ath79/ar71xx_regs.h>
+
+#include "common.h"
+#include "dev-gpio-buttons.h"
+#include "dev-eth.h"
+#include "dev-leds-gpio.h"
+#include "dev-m25p80.h"
+#include "dev-usb.h"
+#include "dev-wmac.h"
+#include "machtypes.h"
+
+#define TL_WR810N_GPIO_SWITCH_B1	0
+#define TL_WR810N_GPIO_SWITCH_B0	1
+#define TL_WR810N_GPIO_USB_POWER	11
+#define TL_WR800N_GPIO_BTN_RESET	12
+#define TL_WR800N_GPIO_LED_SYSTEM	13
+
+#define TL_WR800N_KEYS_POLL_INTERVAL	20 /* msecs */
+#define TL_WR800N_KEYS_DEBOUNCE_INTERVAL (3 * TL_WR800N_KEYS_POLL_INTERVAL)
+
+#define TL_WR800N_WMAC_CALDATA_OFFSET	0x1000
+
+static const char *tl_wr800n_part_probes[] = {
+	"tp-link",
+	NULL,
+};
+
+static struct flash_platform_data tl_wr800n_flash_data = {
+	.part_probes = tl_wr800n_part_probes,
+};
+
+static struct gpio_led tl_wr800n_leds_gpio[] __initdata = {
+	{
+		.name		= "tp-link:blue:system",
+		.gpio		= TL_WR800N_GPIO_LED_SYSTEM,
+		.active_low	= 1,
+	},
+};
+
+static struct gpio_keys_button tl_wr800n_gpio_keys[] __initdata = {
+	{
+		.desc		= "reset",
+		.type		= EV_KEY,
+		.code		= KEY_RESTART,
+		.debounce_interval = TL_WR800N_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR800N_GPIO_BTN_RESET,
+		.active_low	= 1,
+	},
+	{
+		.desc		= "switch_b0",
+		.type		= EV_SW,
+		.code		= BTN_0,
+		.debounce_interval = TL_WR800N_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR800N_GPIO_SWITCH_B0,
+		.active_low	= 0,
+	},
+	{
+		.desc		= "switch_b1",
+		.type		= EV_SW,
+		.code		= BTN_1,
+		.debounce_interval = TL_WR800N_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR800N_GPIO_SWITCH_B1,
+		.active_low	= 0,
+	},
+};
+
+static void __init tl_ap143_setup(int lan_mac_offset)
+{
+	u8 *mac = (u8 *) KSEG1ADDR(0x1f01fc00);
+	u8 *art = (u8 *) KSEG1ADDR(0x1fff0000);
+
+	ath79_register_m25p80(&tl_wr800n_flash_data);
+
+	ath79_setup_ar933x_phy4_switch(false, false);
+
+	ath79_register_mdio(0, 0x0);
+
+	/* WAN */
+	ath79_eth0_data.phy_if_mode = PHY_INTERFACE_MODE_MII;
+	ath79_eth0_data.duplex = DUPLEX_FULL;
+	ath79_eth0_data.speed = SPEED_100;
+	ath79_eth0_data.phy_mask = BIT(4);
+	ath79_init_mac(ath79_eth0_data.mac_addr, mac, 1);
+	ath79_register_eth(0);
+
+	/* LAN */
+	ath79_switch_data.phy4_mii_en = 1;
+	ath79_eth1_data.phy_if_mode = PHY_INTERFACE_MODE_GMII;
+	ath79_eth1_data.duplex = DUPLEX_FULL;
+	ath79_eth1_data.speed = SPEED_1000;
+	ath79_switch_data.phy_poll_mask |= BIT(4);
+	ath79_init_mac(ath79_eth1_data.mac_addr, mac, lan_mac_offset);
+	ath79_register_eth(1);
+
+	ath79_register_wmac(art + TL_WR800N_WMAC_CALDATA_OFFSET, mac);
+
+	ath79_register_leds_gpio(-1,
+				 ARRAY_SIZE(tl_wr800n_leds_gpio),
+				 tl_wr800n_leds_gpio);
+	ath79_register_gpio_keys_polled(-1,
+					TL_WR800N_KEYS_POLL_INTERVAL,
+					ARRAY_SIZE(tl_wr800n_gpio_keys),
+					tl_wr800n_gpio_keys);
+}
+
+static void __init tl_wr810n_setup(void)
+{
+	tl_ap143_setup(-1);
+
+	gpio_request_one(TL_WR810N_GPIO_USB_POWER,
+			 GPIOF_OUT_INIT_HIGH | GPIOF_EXPORT_DIR_FIXED,
+			 "USB power");
+	ath79_register_usb();
+}
+
+static void __init tl_wr800n_v2_setup(void)
+{
+	tl_ap143_setup(0);
+}
+
+MIPS_MACHINE(ATH79_MACH_TL_WR800N, "TL-WR800N", "TP-LINK TL-WR800N",
+	     tl_wr800n_setup);
+
+MIPS_MACHINE(ATH79_MACH_TL_WR800N_V2, "TL-WR800N-v2", "TP-LINK TL-WR800N v2",
+	     tl_wr800n_v2_setup);
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr841n-v8.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr841n-v8.c
index 73cfdd9cc..870ed8ff2 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr841n-v8.c
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr841n-v8.c
@@ -188,7 +188,8 @@ static void __init tl_ap123_setup(void)
 
 	ath79_register_m25p80(&tl_wr841n_v8_flash_data);
 
-	ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_SW_PHY_SWAP);
+	//ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_SW_PHY_SWAP);
+	ath79_setup_ar934x_eth_cfg(AR934X_ETH_CFG_SW_ONLY_MODE);
 
 	ath79_register_mdio(1, 0x0);
 
@@ -197,9 +198,9 @@ static void __init tl_ap123_setup(void)
 
 	/* GMAC0 is connected to the PHY0 of the internal switch */
 	ath79_switch_data.phy4_mii_en = 1;
-	ath79_switch_data.phy_poll_mask = BIT(0);
+	ath79_switch_data.phy_poll_mask = BIT(4);
 	ath79_eth0_data.phy_if_mode = PHY_INTERFACE_MODE_MII;
-	ath79_eth0_data.phy_mask = BIT(0);
+	ath79_eth0_data.phy_mask = BIT(4);
 	ath79_eth0_data.mii_bus_dev = &ath79_mdio1_device.dev;
 	ath79_register_eth(0);
 
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr882n-v1.c b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr882n-v1.c
new file mode 100644
index 000000000..764ad8b34
--- /dev/null
+++ b/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr882n-v1.c
@@ -0,0 +1,143 @@
+/*
+ *  TP-LINK TL-WR882N v1 board support
+ *
+ *  Copyright (C) 2016 David Lutz <kpanic@ff3l.net>
+ * 
+ *  This program is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 as published
+ *  by the Free Software Foundation.
+ *  zcz note: system led not showing in luci,so has to  .active_low=0
+ */
+
+#include <linux/gpio.h>
+#include <linux/platform_device.h>
+
+#include <asm/mach-ath79/ath79.h>
+#include <asm/mach-ath79/ar71xx_regs.h>
+
+#include "common.h"
+#include "dev-eth.h"
+#include "dev-gpio-buttons.h"
+#include "dev-leds-gpio.h"
+#include "dev-m25p80.h"
+#include "dev-wmac.h"
+#include "machtypes.h"
+
+
+#define TL_WR882N_V1_GPIO_LED_WLAN		8
+#define TL_WR882N_V1_GPIO_LED_WAN		2
+#define TL_WR882N_V1_GPIO_LED_LAN1		4
+#define TL_WR882N_V1_GPIO_LED_LAN2		5
+#define TL_WR882N_V1_GPIO_LED_LAN3		6
+#define TL_WR882N_V1_GPIO_LED_LAN4		7
+#define TL_WR882N_V1_GPIO_LED_SYSTEM		18
+#define TL_WR882N_V1_GPIO_BTN_RESET		1
+#define TL_WR882N_V1_GPIO_BTN_RFKILL		16
+
+#define TL_WR882N_KEYS_POLL_INTERVAL		20
+#define TL_WR882N_KEYS_DEBOUNCE_INTERVAL	(3 * TL_WR882N_KEYS_POLL_INTERVAL)
+
+
+static struct gpio_led tl_wr882n_v1_leds_gpio[] __initdata = {
+	{
+		.name		= "tp-link:blue:wlan",
+		.gpio		= TL_WR882N_V1_GPIO_LED_WLAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:blue:wan",
+		.gpio		= TL_WR882N_V1_GPIO_LED_WAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:blue:lan1",
+		.gpio		= TL_WR882N_V1_GPIO_LED_LAN1,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:blue:lan2",
+		.gpio		= TL_WR882N_V1_GPIO_LED_LAN2,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:blue:lan3",
+		.gpio		= TL_WR882N_V1_GPIO_LED_LAN3,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:blue:lan4",
+		.gpio		= TL_WR882N_V1_GPIO_LED_LAN4,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:blue:system",
+		.gpio		= TL_WR882N_V1_GPIO_LED_SYSTEM,
+		.active_low	= 0,
+	},
+};
+
+static struct gpio_keys_button tl_wr882n_v1_gpio_keys[] __initdata = {
+	{
+		.desc		= "Reset button",
+		.type		= EV_KEY,
+		.code		= KEY_RESTART,
+		.debounce_interval = TL_WR882N_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR882N_V1_GPIO_BTN_RESET,
+		.active_low	= 1,
+	}, {
+		.desc		= "RFKILL button",
+		.type		= EV_KEY,
+		.code		= KEY_RFKILL,
+		.debounce_interval = TL_WR882N_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR882N_V1_GPIO_BTN_RFKILL,
+		.active_low	= 0,
+	}
+};
+
+
+static const char *tl_wr882n_part_probes[] = {
+	"tp-link",
+	NULL,
+};
+
+static struct flash_platform_data tl_wr882n_flash_data = {
+	.part_probes	= tl_wr882n_part_probes,
+};
+
+
+static void __init tl_wr882n_setup(void)
+{
+	u8 *mac = (u8 *) KSEG1ADDR(0x1f01fc00);
+	u8 *ee = (u8 *) KSEG1ADDR(0x1fff1000);
+
+	ath79_register_m25p80(&tl_wr882n_flash_data);
+
+	ath79_register_mdio(0, 0x0);
+
+	ath79_init_mac(ath79_eth0_data.mac_addr, mac, 1);
+	ath79_init_mac(ath79_eth1_data.mac_addr, mac, 0);
+
+	ath79_switch_data.phy4_mii_en = 1;
+
+	ath79_register_eth(0);
+	ath79_register_eth(1);
+
+	ath79_register_wmac(ee, mac);
+
+}
+
+static void __init tl_wr882n_v1_setup(void)
+{
+	tl_wr882n_setup();
+
+	ath79_register_leds_gpio(-1, ARRAY_SIZE(tl_wr882n_v1_leds_gpio),
+				 tl_wr882n_v1_leds_gpio);
+
+	ath79_register_gpio_keys_polled(-1, TL_WR882N_KEYS_POLL_INTERVAL,
+					ARRAY_SIZE(tl_wr882n_v1_gpio_keys),
+					tl_wr882n_v1_gpio_keys);
+}
+
+
+MIPS_MACHINE(ATH79_MACH_TL_WR882N_V1, "TL-WR882N-v1", "TP-LINK TL-WR882N v1",
+	     tl_wr882n_v1_setup);
diff --git a/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h b/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
index 4b3f46af2..744d4239a 100644
--- a/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
+++ b/target/linux/ar71xx/files/arch/mips/ath79/machtypes.h
@@ -314,6 +314,7 @@ enum ath79_mach_type {
 	ATH79_MACH_TL_WR941ND,			/* TP-LINK TL-WR941ND */
 	ATH79_MACH_TL_WR941ND_V5,		/* TP-LINK TL-WR941ND v5 */
 	ATH79_MACH_TL_WR941ND_V6,		/* TP-LINK TL-WR941ND v6 */
+	ATH79_MACH_TL_WR882N_V1,		/* TP-LINK TL-WR882N v1 */
 	ATH79_MACH_TL_WR940N_V4,		/* TP-LINK TL-WR940N v4 */
 	ATH79_MACH_TL_WR940N_V6,		/* TP-LINK TL-WR940N v6 */
 	ATH79_MACH_TL_WR942N_V1,		/* TP-LINK TL-WR942N v1 */
diff --git a/target/linux/ar71xx/generic/config-default b/target/linux/ar71xx/generic/config-default
index 25b58ae91..91b71dfc4 100644
--- a/target/linux/ar71xx/generic/config-default
+++ b/target/linux/ar71xx/generic/config-default
@@ -169,6 +169,7 @@ CONFIG_ATH79_MACH_TL_WR810N=y
 CONFIG_ATH79_MACH_TL_WR810N_V2=y
 CONFIG_ATH79_MACH_TL_WR841N_V8=y
 CONFIG_ATH79_MACH_TL_WR841N_V9=y
+CONFIG_ATH79_MACH_TL_WR882N_V1=y
 CONFIG_ATH79_MACH_TL_WR902AC_V1=y
 CONFIG_ATH79_MACH_TL_WR942N_V1=y
 CONFIG_ATH79_MACH_TS_D084=y
diff --git a/target/linux/ar71xx/image/generic-tp-link.mk b/target/linux/ar71xx/image/generic-tp-link.mk
index dac5ad285..8e2ec9f77 100644
--- a/target/linux/ar71xx/image/generic-tp-link.mk
+++ b/target/linux/ar71xx/image/generic-tp-link.mk
@@ -482,6 +482,15 @@ define Device/tl-wr810n-v2
 endef
 TARGET_DEVICES += tl-wr810n-v2
 
+define Device/tl-wr882n-v1
+  $(Device/tplink-16mlzma)
+  DEVICE_TITLE := TP-LINK TL-WR882N v1
+  BOARDNAME := TL-WR882N-v1
+  DEVICE_PROFILE := TLWR882
+  TPLINK_HWID := 0x08820001
+endef
+TARGET_DEVICES += tl-wr882n-v1
+
 define Device/tl-wr842n-v1
   $(Device/tplink-8m)
   DEVICE_TITLE := TP-LINK TL-WR842N/ND v1
diff --git a/target/linux/ar71xx/image/tiny-tp-link.mk b/target/linux/ar71xx/image/tiny-tp-link.mk
index 5874808d2..f526dd697 100644
--- a/target/linux/ar71xx/image/tiny-tp-link.mk
+++ b/target/linux/ar71xx/image/tiny-tp-link.mk
@@ -437,7 +437,7 @@ endef
 TARGET_DEVICES += tl-wr741nd-v2
 
 define Device/tl-wr741nd-v4
-  $(Device/tplink-4mlzma)
+  $(Device/tplink-8mlzma)
   DEVICE_TITLE := TP-LINK TL-WR741N/ND v4
   BOARDNAME := TL-WR741ND-v4
   DEVICE_PROFILE := TLWR741
@@ -551,7 +551,7 @@ endef
 TARGET_DEVICES += tl-wr841-v7
 
 define Device/tl-wr841-v8
-  $(Device/tplink-4mlzma)
+  $(Device/tplink-8mlzma)
   DEVICE_TITLE := TP-LINK TL-WR841N/ND v8
   BOARDNAME := TL-WR841N-v8
   DEVICE_PROFILE := TLWR841
@@ -560,7 +560,7 @@ endef
 TARGET_DEVICES += tl-wr841-v8
 
 define Device/tl-wr841-v9
-  $(Device/tplink-4mlzma)
+  $(Device/tplink-8mlzma)
   DEVICE_TITLE := TP-LINK TL-WR841N/ND v9
   BOARDNAME := TL-WR841N-v9
   DEVICE_PROFILE := TLWR841
@@ -576,7 +576,7 @@ endef
 TARGET_DEVICES += tl-wr841-v10
 
 define Device/tl-wr841-v11
-  $(Device/tplink-4mlzma)
+  $(Device/tplink-8mlzma)
   DEVICE_TITLE := TP-LINK TL-WR841N/ND v11
   BOARDNAME := TL-WR841N-v11
   DEVICE_PROFILE := TLWR841
@@ -612,6 +612,15 @@ define Device/tl-wr847n-v8
 endef
 TARGET_DEVICES += tl-wr847n-v8
 
+define Device/tl-wr882n-v1
+  $(Device/tplink-8mlzma)
+  DEVICE_TITLE := TP-LINK TL-WR882N v1
+  BOARDNAME := TL-WR882N-v1
+  DEVICE_PROFILE := TLWR882
+  TPLINK_HWID := 0x08820001
+endef
+TARGET_DEVICES += tl-wr882n-v1
+
 define Device/tl-wr940n-v4
   $(Device/tplink-4mlzma)
   DEVICE_TITLE := TP-LINK TL-WR940N v4
@@ -693,3 +702,12 @@ define Device/tl-wr941nd-v6-cn
   TPLINK_HWID := 0x09410006
 endef
 TARGET_DEVICES += tl-wr941nd-v6-cn
+
+define Device/tl-wr800n-v2
+  $(Device/tplink-8mlzma)
+  DEVICE_TITLE := TP-LINK TL-WR800N v2
+  BOARDNAME := TL-WR800N-v2
+  DEVICE_PROFILE := TLWR800
+  TPLINK_HWID := 0x08000002
+endef
+TARGET_DEVICES += tl-wr800n-v2
diff --git a/target/linux/ar71xx/tiny/config-default b/target/linux/ar71xx/tiny/config-default
index 1148d8c45..06913e393 100644
--- a/target/linux/ar71xx/tiny/config-default
+++ b/target/linux/ar71xx/tiny/config-default
@@ -44,6 +44,7 @@ CONFIG_ATH79_MACH_TL_WR840N_V2=y
 CONFIG_ATH79_MACH_TL_WR841N_V1=y
 CONFIG_ATH79_MACH_TL_WR841N_V8=y
 CONFIG_ATH79_MACH_TL_WR841N_V9=y
+CONFIG_ATH79_MACH_TL_WR882N_V1=y
 CONFIG_ATH79_MACH_TL_WR940N_V4=y
 CONFIG_ATH79_MACH_TL_WR941ND=y
 CONFIG_ATH79_MACH_TL_WR941ND_V6=y
diff --git a/target/linux/ath79/dts/ar7240_tplink.dtsi b/target/linux/ath79/dts/ar7240_tplink.dtsi
index 7e54cd46b..945b9707a 100644
--- a/target/linux/ath79/dts/ar7240_tplink.dtsi
+++ b/target/linux/ath79/dts/ar7240_tplink.dtsi
@@ -83,12 +83,12 @@
 
 			partition@20000 {
 				compatible = "tplink,firmware";
-				reg = <0x20000 0x3d0000>;
+				reg = <0x20000 0x7d0000>;
 				label = "firmware";
 			};
 
-			art: partition@3f0000 {
-				reg = <0x3f0000 0x10000>;
+			art: partition@7f0000 {
+				reg = <0x7f0000 0x10000>;
 				label = "art";
 				read-only;
 			};
diff --git a/target/linux/ath79/dts/ar7241_tplink.dtsi b/target/linux/ath79/dts/ar7241_tplink.dtsi
index 3cbedc3bf..850bcf8f7 100644
--- a/target/linux/ath79/dts/ar7241_tplink.dtsi
+++ b/target/linux/ath79/dts/ar7241_tplink.dtsi
@@ -70,12 +70,12 @@
 
 			partition@20000 {
 				compatible = "tplink,firmware";
-				reg = <0x20000 0x3d0000>;
+				reg = <0x20000 0x7d0000>;
 				label = "firmware";
 			};
 
-			partition@3f0000 {
-				reg = <0x3f0000 0x10000>;
+			partition@7f0000 {
+				reg = <0x7f0000 0x10000>;
 				label = "art";
 				read-only;
 			};
diff --git a/target/linux/ath79/dts/ar9341_tplink.dtsi b/target/linux/ath79/dts/ar9341_tplink.dtsi
index dc97b103a..f32775b41 100644
--- a/target/linux/ath79/dts/ar9341_tplink.dtsi
+++ b/target/linux/ath79/dts/ar9341_tplink.dtsi
@@ -83,6 +83,7 @@
 
 &eth0 {
 	status = "okay";
+	phy-mode = "mii";
 
 	phy-handle = <&swphy0>;
 
diff --git a/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8.dts b/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8.dts
index 9f2e680b6..13fa745fa 100644
--- a/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8.dts
+++ b/target/linux/ath79/dts/ar9341_tplink_tl-wr841-v8.dts
@@ -50,14 +50,42 @@
 			partition@20000 {
 				compatible = "tplink,firmware";
 				label = "firmware";
-				reg = <0x020000 0x3d0000>;
+				reg = <0x020000 0x7d0000>;
 			};
 
-			art: partition@3f0000 {
+			art: partition@7f0000 {
 				label = "art";
-				reg = <0x3f0000 0x010000>;
+				reg = <0x7f0000 0x010000>;
 				read-only;
 			};
 		};
 	};
 };
+
+&eth0 {
+        status = "okay";
+
+        phy-handle = <&swphy0>;
+        mtd-mac-address = <&uboot 0x1fc00>;
+        mtd-mac-address-increment = <(-1)>;
+};
+
+&eth1 {
+        status = "okay";
+
+        mtd-mac-address = <&uboot 0x1fc00>;
+
+        pll-data = <0x06000000 0x00000101 0x00001616>;
+
+        gmac-config {
+                device = <&gmac>;
+                switch-phy-swap = <1>;
+        };
+};
+
+&wmac {
+        status = "okay";
+
+        mtd-cal-data = <&art 0x1000>;
+        mtd-mac-address = <&uboot 0x1fc00>;
+};
diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr800n-v2.dts b/target/linux/ath79/dts/qca9533_tplink_tl-wr800n-v2.dts
new file mode 100644
index 000000000..8084c1148
--- /dev/null
+++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr800n-v2.dts
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+//add bye zcz
+
+#include "qca953x_tplink_tl-wr800n_16M.dtsi"
+
+/ {
+	compatible = "tplink,tl-wr800n-v2", "qca,qca9533";
+	model = "TP-Link TL-WR800N v2";
+};
diff --git a/target/linux/ath79/dts/qca9533_tplink_tl-wr841.dtsi b/target/linux/ath79/dts/qca9533_tplink_tl-wr841.dtsi
index e85f37f42..a43d45d10 100644
--- a/target/linux/ath79/dts/qca9533_tplink_tl-wr841.dtsi
+++ b/target/linux/ath79/dts/qca9533_tplink_tl-wr841.dtsi
@@ -97,12 +97,12 @@
 			partition@20000 {
 				compatible = "tplink,firmware";
 				label = "firmware";
-				reg = <0x020000 0x3d0000>;
+				reg = <0x020000 0x7d0000>;
 			};
 
-			art: partition@3f0000 {
+			art: partition@7f0000 {
 				label = "art";
-				reg = <0x3f0000 0x010000>;
+				reg = <0x7f0000 0x010000>;
 				read-only;
 			};
 		};
diff --git a/target/linux/ath79/dts/qca953x_tplink_tl-wr800n_16M.dtsi b/target/linux/ath79/dts/qca953x_tplink_tl-wr800n_16M.dtsi
new file mode 100644
index 000000000..356598385
--- /dev/null
+++ b/target/linux/ath79/dts/qca953x_tplink_tl-wr800n_16M.dtsi
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+//add bye zcz
+
+#include "qca953x.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	aliases {
+		led-boot = &led_system;
+		led-failsafe = &led_system;
+		led-running = &led_system;
+		led-upgrade = &led_system;
+		label-mac-device = &eth1;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_system: system {
+			label = "tp-link:blue:system";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		switch_b0 {
+			label = "switch_b0";
+			linux,input-type = <EV_SW>;
+			linux,code = <BTN_0>;
+			gpios = <&gpio 1 GPIO_ACTIVE_HIGH>;
+			debounce-interval = <60>;
+		};
+
+		switch_b1 {
+			label = "switch_b1";
+			linux,input-type = <EV_SW>;
+			linux,code = <BTN_1>;
+			gpios = <&gpio 0 GPIO_ACTIVE_HIGH>;
+			debounce-interval = <60>;
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+//reg = <begining end>
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot: partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x020000>;
+				read-only;
+			};
+
+			partition@20000 {
+				compatible = "tplink,firmware";
+				label = "firmware";
+				reg = <0x020000 0xfd0000>;
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x010000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	phy-handle = <&swphy4>;
+
+	mtd-mac-address = <&uboot 0x1fc00>;
+	mtd-mac-address-increment = <1>;
+};
+
+&eth1 {
+	mtd-mac-address = <&uboot 0x1fc00>;
+};
+
+&wmac {
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&uboot 0x1fc00>;
+};
diff --git a/target/linux/ath79/dts/tp9343_tplink_tl-wr882n-v1.dts b/target/linux/ath79/dts/tp9343_tplink_tl-wr882n-v1.dts
new file mode 100644
index 000000000..267f480d5
--- /dev/null
+++ b/target/linux/ath79/dts/tp9343_tplink_tl-wr882n-v1.dts
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "tp9343_tplink_tl-wr94x.dtsi"
+
+/ {
+	compatible = "tplink,tl-wr882-v1", "qca,tp9343";
+	model = "TP-Link TL-WR882N v1";
+
+	aliases {
+		led-boot = &led_system;
+		led-failsafe = &led_system;
+		led-running = &led_system;
+		led-upgrade = &led_system;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_system: system {
+			label = "tp-link:blue:system";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		wan_blue {
+			label = "tp-link:blue:wan";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+
+		lan1 {
+			label = "tp-link:blue:lan1";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+		};
+
+		lan2 {
+			label = "tp-link:blue:lan2";
+			gpios = <&gpio 5 GPIO_ACTIVE_LOW>;
+		};
+
+		lan3 {
+			label = "tp-link:blue:lan3";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+		};
+
+		lan4 {
+			label = "tp-link:blue:lan4";
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&eth1 {
+	mtd-mac-address = <&uboot 0x1fc00>;
+//	mtd-mac-address-increment = <(-1)>;
+};
diff --git a/target/linux/ath79/dts/tp9343_tplink_tl-wx.dtsi b/target/linux/ath79/dts/tp9343_tplink_tl-wx.dtsi
index 30e2cb643..06fe7783e 100644
--- a/target/linux/ath79/dts/tp9343_tplink_tl-wx.dtsi
+++ b/target/linux/ath79/dts/tp9343_tplink_tl-wx.dtsi
@@ -35,12 +35,12 @@
 			partition@20000 {
 				compatible = "tplink,firmware";
 				label = "firmware";
-				reg = <0x020000 0x3d0000>;
+				reg = <0x020000 0x7d0000>;
 			};
 
-			art: partition@3f0000 {
+			art: partition@7f0000 {
 				label = "art";
-				reg = <0x3f0000 0x010000>;
+				reg = <0x7f0000 0x010000>;
 				read-only;
 			};
 		};
diff --git a/target/linux/ath79/generic/base-files/etc/board.d/01_leds b/target/linux/ath79/generic/base-files/etc/board.d/01_leds
index ef7a0d4e3..7dff7d443 100755
--- a/target/linux/ath79/generic/base-files/etc/board.d/01_leds
+++ b/target/linux/ath79/generic/base-files/etc/board.d/01_leds
@@ -235,6 +235,7 @@ tplink,archer-c6-v2-us)
 	ucidef_set_led_switch "wan" "WAN" "tp-link:green:wan" "switch0" "0x02"
 	;;
 tplink,archer-c25-v1|\
+tplink,tl-wr882n-v1|\
 tplink,tl-wr842n-v3)
 	ucidef_set_led_netdev "wan" "WAN" "tp-link:green:wan" "eth1"
 	ucidef_set_led_switch "lan1" "LAN1" "tp-link:green:lan1" "switch0" "0x10"
diff --git a/target/linux/ath79/generic/base-files/etc/board.d/02_network b/target/linux/ath79/generic/base-files/etc/board.d/02_network
index 518ca6f91..58e206a02 100755
--- a/target/linux/ath79/generic/base-files/etc/board.d/02_network
+++ b/target/linux/ath79/generic/base-files/etc/board.d/02_network
@@ -103,6 +103,7 @@ ath79_setup_interfaces()
 	tplink,tl-wdr3500-v1|\
 	tplink,tl-wr842n-v1|\
 	tplink,tl-wr842n-v3|\
+	tplink,tl-wr882n-v1|\
 	ubnt,airrouter)
 		ucidef_set_interface_wan "eth1"
 		ucidef_add_switch "switch0" \
diff --git a/target/linux/ath79/image/tiny-tp-link.mk b/target/linux/ath79/image/tiny-tp-link.mk
index c918c6baa..76d139266 100644
--- a/target/linux/ath79/image/tiny-tp-link.mk
+++ b/target/linux/ath79/image/tiny-tp-link.mk
@@ -348,7 +348,7 @@ endef
 TARGET_DEVICES += tplink_tl-wr841-v7
 
 define Device/tplink_tl-wr841-v8
-  $(Device/tplink-4mlzma)
+  $(Device/tplink-8mlzma)
   SOC := ar9341
   DEVICE_MODEL := TL-WR841N/ND
   DEVICE_VARIANT := v8
@@ -413,6 +413,16 @@ define Device/tplink_tl-wr940n-v3
 endef
 TARGET_DEVICES += tplink_tl-wr940n-v3
 
+define Device/tplink_tl-wr882n-v1
+  $(Device/tplink-8mlzma)
+  SOC := tp9343
+  DEVICE_MODEL := TL-WR882N
+  DEVICE_VARIANT := v1
+  TPLINK_HWID := 0x08820001
+  SUPPORTED_DEVICES += tl-wr882n-v1
+endef
+TARGET_DEVICES += tplink_tl-wr882n-v1
+
 define Device/tplink_tl-wr940n-v4
   $(Device/tplink-4mlzma)
   SOC := tp9343
@@ -428,7 +438,7 @@ endef
 TARGET_DEVICES += tplink_tl-wr940n-v4
 
 define Device/tplink_tl-wr940n-v6
-  $(Device/tplink-4mlzma)
+  $(Device/tplink-8mlzma)
   SOC := tp9343
   DEVICE_MODEL := TL-WR940N
   DEVICE_VARIANT := v6
@@ -456,7 +466,7 @@ endef
 TARGET_DEVICES += tplink_tl-wr941-v2
 
 define Device/tplink_tl-wr941-v4
-  $(Device/tplink-4m)
+  $(Device/tplink-8m)
   SOC := ar7240
   DEVICE_MODEL := TL-WR941ND
   DEVICE_VARIANT := v4
@@ -486,3 +496,13 @@ define Device/tplink_tl-wr941n-v7-cn
   TPLINK_HWID := 0x09410007
 endef
 TARGET_DEVICES += tplink_tl-wr941n-v7-cn
+
+define Device/tplink_tl-wr800n-v2
+  $(Device/tplink-16mlzma)
+  SOC := qca9533
+  DEVICE_MODEL := TL-WR800N
+  DEVICE_VARIANT := v2
+  TPLINK_HWID := 0x08000002
+endef
+TARGET_DEVICES += tplink_tl-wr800n-v2
+
diff --git a/target/linux/ath79/tiny/base-files/etc/board.d/01_leds b/target/linux/ath79/tiny/base-files/etc/board.d/01_leds
index bc0b15bba..ee14af62d 100755
--- a/target/linux/ath79/tiny/base-files/etc/board.d/01_leds
+++ b/target/linux/ath79/tiny/base-files/etc/board.d/01_leds
@@ -101,6 +101,7 @@ tplink,tl-wr941-v4)
 	;;
 tplink,tl-wr940n-v3|\
 tplink,tl-wr940n-v4|\
+tplink,tl-wr882n-v1|\
 tplink,tl-wr941nd-v6)
 	ucidef_set_led_netdev "wan" "WAN" "tp-link:blue:wan" "eth1"
 	ucidef_set_led_switch "lan1" "LAN1" "tp-link:blue:lan1" "switch0" "0x10"
diff --git a/target/linux/ath79/tiny/base-files/etc/board.d/02_network b/target/linux/ath79/tiny/base-files/etc/board.d/02_network
index cbcdbe9a8..dcad49af9 100755
--- a/target/linux/ath79/tiny/base-files/etc/board.d/02_network
+++ b/target/linux/ath79/tiny/base-files/etc/board.d/02_network
@@ -60,12 +60,13 @@ ath79_setup_interfaces()
 	tplink,tl-wr841-v11|\
 	tplink,tl-wr841-v12|\
 	tplink,tl-wr940n-v3|\
+	tplink,tl-wr882n-v1|\
 	tplink,tl-wr940n-v4|\
 	tplink,tl-wr940n-v6|\
 	tplink,tl-wr941nd-v6)
 		ucidef_set_interface_wan "eth1"
 		ucidef_add_switch "switch0" \
-			"0@eth0" "1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1"
+			"0@eth0" "1:lan:1" "2:lan:2" "3:lan:3" "4:lan:4"
 		;;
 	tplink,tl-mr3420-v2|\
 	tplink,tl-wr740n-v4|\
diff --git a/tools/firmware-utils/src/mktplinkfw.c b/tools/firmware-utils/src/mktplinkfw.c
index ed785b48e..502d1d2ca 100644
--- a/tools/firmware-utils/src/mktplinkfw.c
+++ b/tools/firmware-utils/src/mktplinkfw.c
@@ -128,7 +128,7 @@ static struct flash_layout layouts[] = {
 		.rootfs_ofs	= 0x140000,
 	}, {
 		.id		= "4Mlzma",
-		.fw_max_len	= 0x3c0000,
+		.fw_max_len	= 0x7c0000,
 		.kernel_la	= 0x80060000,
 		.kernel_ep	= 0x80060000,
 		.rootfs_ofs	= 0x100000,
